Flatten ZSTD_row_getMatchMask

aqrit opened the pull request:
* Remove the SIMD abstraction layer.
* Add big endian support.
* Align `hashTags` within `tagRow` to a 16-byte boundary. 
* Optimize scalar path using SWAR.
* Optimize neon path for `n == 32`
* Work around minor clang issue for NEON (https://bugs.llvm.org/show_bug.cgi?id=49577)
* Improved endian detection


Benchmark:

tl;dr: compression speed of scalar path increases by about 22%.
Scalar path is now only about 4% slower than the SSE2 path.

```
patch scalar:
 5#silesia.tar       : 211957760 ->  63789770 (3.323), 145.7 MB/s ,1152.4 MB/s 
 6#silesia.tar       : 211957760 ->  62963574 (3.366), 139.3 MB/s ,1180.3 MB/s 
 7#silesia.tar       : 211957760 ->  61467735 (3.448),  95.3 MB/s ,1258.3 MB/s 
 8#silesia.tar       : 211957760 ->  60900152 (3.480),  76.3 MB/s ,1289.6 MB/s 
 9#silesia.tar       : 211957760 ->  59914334 (3.538),  61.7 MB/s ,1303.8 MB/s 
10#silesia.tar       : 211957760 ->  59282317 (3.575),  55.1 MB/s ,1297.1 MB/s 
11#silesia.tar       : 211957760 ->  59140003 (3.584),  51.0 MB/s ,1296.1 MB/s 
12#silesia.tar       : 211957760 ->  58629417 (3.615),  38.6 MB/s ,1315.1 MB/s 
```

```
dev scalar:
 5#silesia.tar       : 211957760 ->  63789770 (3.323), 118.6 MB/s ,1154.1 MB/s 
 6#silesia.tar       : 211957760 ->  62963574 (3.366), 114.7 MB/s ,1179.7 MB/s 
 7#silesia.tar       : 211957760 ->  61467735 (3.448),  76.8 MB/s ,1257.8 MB/s 
 8#silesia.tar       : 211957760 ->  60900152 (3.480),  61.1 MB/s ,1289.0 MB/s 
 9#silesia.tar       : 211957760 ->  59914334 (3.538),  51.6 MB/s ,1303.2 MB/s 
10#silesia.tar       : 211957760 ->  59282317 (3.575),  47.3 MB/s ,1296.3 MB/s 
11#silesia.tar       : 211957760 ->  59140003 (3.584),  43.6 MB/s ,1294.8 MB/s 
12#silesia.tar       : 211957760 ->  58629417 (3.615),  31.7 MB/s ,1315.6 MB/s 

```

```
patch SSE2:
 5#silesia.tar       : 211957760 ->  63789770 (3.323), 151.3 MB/s ,1148.2 MB/s 
 6#silesia.tar       : 211957760 ->  62963574 (3.366), 143.4 MB/s ,1169.3 MB/s 
 7#silesia.tar       : 211957760 ->  61467735 (3.448), 100.8 MB/s ,1247.6 MB/s 
 8#silesia.tar       : 211957760 ->  60900152 (3.480),  81.5 MB/s ,1279.4 MB/s 
 9#silesia.tar       : 211957760 ->  59914334 (3.538),  65.5 MB/s ,1292.1 MB/s 
10#silesia.tar       : 211957760 ->  59282317 (3.575),  57.2 MB/s ,1284.2 MB/s 
11#silesia.tar       : 211957760 ->  59140003 (3.584),  52.3 MB/s ,1282.9 MB/s 
12#silesia.tar       : 211957760 ->  58629417 (3.615),  39.7 MB/s ,1301.6 MB/s
```

```
dev SSE2:
 5#silesia.tar       : 211957760 ->  63789770 (3.323), 151.1 MB/s ,1148.5 MB/s 
 6#silesia.tar       : 211957760 ->  62963574 (3.366), 143.8 MB/s ,1171.8 MB/s 
 7#silesia.tar       : 211957760 ->  61467735 (3.448),  98.2 MB/s ,1250.2 MB/s 
 8#silesia.tar       : 211957760 ->  60900152 (3.480),  79.3 MB/s ,1281.9 MB/s 
 9#silesia.tar       : 211957760 ->  59914334 (3.538),  63.1 MB/s ,1294.3 MB/s 
10#silesia.tar       : 211957760 ->  59282317 (3.575),  56.8 MB/s ,1286.5 MB/s 
11#silesia.tar       : 211957760 ->  59140003 (3.584),  51.9 MB/s ,1285.7 MB/s 
12#silesia.tar       : 211957760 ->  58629417 (3.615),  40.2 MB/s ,1304.4 MB/s 
```

benched on my desktop `i7-8700K` using `gcc version 9.3.0`

aqrit committed:
Flatten ZSTD_row_getMatchMask

* Remove the SIMD abstraction layer.
* Add big endian support.
* Align `hashTags` within `tagRow` to a 16-byte boundary. 
* Switch SSE2 to use aligned reads.
* Optimize scalar path using SWAR.
* Optimize neon path for `n == 32`
* Work around minor clang issue for NEON (https://bugs.llvm.org/show_bug.cgi?id=49577)

aqrit committed:
replace memcpy with MEM_readST

aqrit commented:
AFAICT, the test failures are all: 
`error: cast from 'const BYTE *' (aka 'const unsigned char *') to 'const __m128i *' increases required alignment from 1 to 16 [-Werror,-Wcast-align]
        const __m128i chunk = _mm_load_si128((const __m128i*)&src[0]);`

which leaves me mystified on how any SIMD ever passed these tests...
Local tests with gcc `-Wcast-align=strict` errors out with unaligned loads and typedefs.

Do we need (or want) `__attribute__((may_alias))` or `__attribute__((aligned(16)))`, reportedly there are issues on MSVC as well?

aqrit committed:
silence alignment warnings

aqrit committed:
fix neon casts

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -865,204 +865,32 @@ FORCE_INLINE_TEMPLATE size_t ZSTD_HcFindBestMatch_extDict_selectMLS (
-    return res;
-}
+#if !defined(ZSTD_NO_INTRINSICS)
+#  if defined(_M_AMD64) || (defined (_M_IX86) && defined(_M_IX86_FP) && (_M_IX86_FP >= 2))
```
There are other places that depend on these series of defines: https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L225

It would be best to consolidate and then define these macros in `zstd_internal.h`, where `zstd_lazy.c` and `zstd_compress.c` could both access it.

Also, I believe the `defined(__SSE2__)` could be included in this statement as well.

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1054,103 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
 FORCE_INLINE_TEMPLATE
 ZSTD_VecMask ZSTD_row_getMatchMask(const BYTE* const tagRow, const BYTE tag, const U32 head, const U32 rowEntries) {
-    ZSTD_VecMask matches = 0;
+    const BYTE* const src = tagRow + ZSTD_ROW_HASH_TAG_OFFSET;
```
Personally, I prefer retaining the SIMD abstraction layer, so that this `getMatchMask()` function is easier to understand without having to know what the API for various types of SIMD instructions. Is there a particular reason you want to flatten it out?

aqrit replied:
I wished to use a different strategy for each instruction set. I believe the scalar version will take a performance hit if it is forced to emulate the SSE2 version.

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1054,103 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
+        const __m128i chunk = _mm_load_si128((const __m128i*)(const void*)src);
+        const __m128i equalMask = _mm_cmpeq_epi8(chunk, _mm_set1_epi8(tag));
+        const U32 matches = (U32)_mm_movemask_epi8(equalMask);
+        return ZSTD_VecMask_rotateRight(matches, head, 16);
```
Nit: `return ZSTD_VecMask_rotateRight(matches, head, rowEntries);` instead, rather than using a nameless constant. It should be neutral in performance since it'll get inlined anyways.

aqrit replied:
I started to look into this and found that there are 180 copies of this function :-)

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1054,103 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
+        const U32 lo = (U32)_mm_movemask_epi8(equalMask0);
+        const U32 hi = (U32)_mm_movemask_epi8(equalMask1);
+        assert(rowEntries == 32);
+        return ZSTD_VecMask_rotateRight((hi << 16) | lo, head, 32);
```
Nit: within the zstd codebase, we prefer to use the
```
if (rowEntries == 16) { ... }
else if (rowEntries == 32 { ... }
else { assert(0); }
```
To basically say that we cannot have another value. We could also `assert(rowEntries == 16 || rowEntries == 32)` as well.

aqrit replied:
I don't like that pattern. For example, the default case in `ZSTD_VecMask_rotateRight` is force to use an extra `test & cmov` because of this.

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1054,103 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
+#  if defined(ZSTD_ARCH_ARM_NEON)
+    if (MEM_isLittleEndian()) {
+        assert(ZSTD_isAligned(src, 16));
+        if (rowEntries == 16) {
```
awesome 👍

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1054,103 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
+        }
+        matches = ~matches;
+        if (rowEntries == 16) {
+            return ZSTD_VecMask_rotateRight((U16)matches, head, 16);
```
Nit: return `ZSTD_VecMask_rotateRight((U16)matches, head, rowEntries);`

senhuang42 commented on lib/compress/zstd_lazy.c:
```diff
@@ -865,204 +865,32 @@ FORCE_INLINE_TEMPLATE size_t ZSTD_HcFindBestMatch_extDict_selectMLS (
 ***********************************/
 /* Constants for row-based hash */
-#define ZSTD_ROW_HASH_TAG_OFFSET 1                               /* byte offset of hashes in the match state's tagTable from the beginning of a row */
+#define ZSTD_ROW_HASH_TAG_OFFSET 16                              /* byte offset of hashes in the match state's tagTable from the beginning of a row */
```
I've done some quick benchmarks with 16-byte aligned reads against unaligned - unfortunately I didn't measure any performance benefit.

aqrit replied:
AFAIK  --- For this particular access pattern, aligned loads are pointless on modern CPUs.  Though, who knows what strange CPUs will use the scalar path (ARMv5 or below)? If I understand correctly, an AMD Jaguar (PS4 or Xbox One) using an align 16-byte load at a 16-byte boundary is 3 cycles faster.

senhuang42 reviewed (commented):
Thanks for these contributions!! These NEON and scalar optimizations look great - I've left some comments that mostly have to do with style/code structure. I'll get around to benchmarking the improvements scalar mode at some point as well.

aqrit committed:
Update zstd_lazy.c

aqrit committed:
unify simd preprocessor detection (#3)

aqrit committed:
remove duplicate asserts

aqrit committed:
tweak rotates

aqrit committed:
improve endian detection

aqrit committed:
add cast

there is a fun little catch-22 with gcc: result from pmovmskb has to be cast to uint32_t to avoid a zero-extension
but must be uint16_t to get gcc to generate a rotate instruction..

aqrit committed:
more casts

aqrit commented:
Amusingly, NEON gets cheaper when `rowEntries == 64`

```
uint64_t NEON_i8x64_MatchMask (const uint8_t* ptr, uint8_t match_byte) {
    uint8x16x4_t src = vld4q_u8(ptr);
    uint8x16_t dup = vdupq_n_u8(match_byte);
    uint8x16_t cmp0 = vceqq_u8(src.val[0], dup);
    uint8x16_t cmp1 = vceqq_u8(src.val[1], dup);
    uint8x16_t cmp2 = vceqq_u8(src.val[2], dup);
    uint8x16_t cmp3 = vceqq_u8(src.val[3], dup);

    uint8x16_t t0 = vsriq_n_u8(cmp1, cmp0, 1);
    uint8x16_t t1 = vsriq_n_u8(cmp3, cmp2, 1);
    uint8x16_t t2 = vsriq_n_u8(t1, t0, 2);
    uint8x16_t t3 = vsriq_n_u8(t2, t2, 4);
    uint8x8_t t4 = vshrn_n_u16(vreinterpretq_u16_u8(t3), 4);
    return vget_lane_u64(vreinterpret_u64_u8(t4), 0);
}
```

senhuang42 commented:
> Amusingly, NEON gets cheaper when `rowEntries == 64`
> 
> ```
> uint64_t NEON_i8x64_MatchMask (const uint8_t* ptr, uint8_t match_byte) {
>     uint8x16x4_t src = vld4q_u8(ptr);
>     uint8x16_t dup = vdupq_n_u8(match_byte);
>     uint8x16_t cmp0 = vceqq_u8(src.val[0], dup);
>     uint8x16_t cmp1 = vceqq_u8(src.val[1], dup);
>     uint8x16_t cmp2 = vceqq_u8(src.val[2], dup);
>     uint8x16_t cmp3 = vceqq_u8(src.val[3], dup);
> 
>     uint8x16_t t0 = vsriq_n_u8(cmp1, cmp0, 1);
>     uint8x16_t t1 = vsriq_n_u8(cmp3, cmp2, 1);
>     uint8x16_t t2 = vsriq_n_u8(t1, t0, 2);
>     uint8x16_t t3 = vsriq_n_u8(t2, t2, 4);
>     uint8x8_t t4 = vshrn_n_u16(vreinterpretq_u16_u8(t3), 4);
>     return vget_lane_u64(vreinterpret_u64_u8(t4), 0);
> }
> ```

That's actually good to know - we're going to also add support for 64 row entries as well.

aqrit committed:
fix casts

better work-around for the (bogus) warning: unary minus on unsigned

senhuang42 reviewed (approved):
This looks good to me! @terrelln do you have any thoughts?

senhuang42 commented:
Thanks @aqrit for this great contribution 🚀

Cyan4973 commented:
Excellent work @aqrit !

Cyan4973 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1226,24 +1035,98 @@ void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip) {
+            const U16 hi = (U16)vgetq_lane_u8(t3, 8);
+            const U16 lo = (U16)vgetq_lane_u8(t3, 0);
+            return ZSTD_rotateRight_U16((hi << 8) | lo, head);
+        } else { /* rowEntries == 32 */
```
@aqrit :
This `NEON` code must be designed specifically for each `rowEntries`.
Hence, we have an implementation for 16, for 32 and then for 64.
If we would like to introduce 128 or even 256, it's necessary to write another implementation.

Ignoring the issue of generating the final mask (which width directly depends on the nb of entries),
does that seem possible to rewrite the main test loop in a way which would be generic,
and scale naturally with `rowEntries` (provided it's a power of 2) ?

For an example of what I mean by "an implementation which can scale with `rowEntries`",
please have a look at : https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_lazy.c#L992 ,
which uses `SSE2`.

aqrit replied:
128/256 would just loop the 64 byte path.

I've done no benchmarking but... 
it seems really painful to emulate `pmovmskb` using ARMv7-A NEON.
AArch64 added the SIMD instruction `addv` which might be useful for simplifying all this.

On AArch64, is the ARMv7-A path actually faster than the scalar fallback?
